src/Microsoft.AspNetCore.Razor.Language/Legacy/CSharpCodeParser.cs

```diff
@@ -1587,165 +1587,191 @@ private void EnsureDirectiveIsAtStartOfLine()
-                            CurrentSymbol.Content.Length);
-                    }
-
-                    Output(SpanKindInternal.MetaCode, AcceptedCharactersInternal.WhiteSpace);
-                    break;
-                case DirectiveKind.RazorBlock:
-                    AcceptWhile(IsSpacingToken(includeNewLines: true, includeComments: true));
-                    Output(SpanKindInternal.Markup, AcceptedCharactersInternal.AllWhiteSpace);
+                            IsNested = wasNested;
 
-                    ParseDirectiveBlock(descriptor, parseChildren: (startingBraceLocation) =>
-                    {
-                        // When transitioning to the HTML parser we no longer want to act as if we're in a nested C# state.
-                        // For instance, if <div>@hello.</div> is in a nested C# block we don't want the trailing '.' to be handled
-                        // as C#; it should be handled as a period because it's wrapped in markup.
-                        var wasNested = IsNested;
-                        IsNested = false;
+                            NextToken();
+                        });
+                        break;
+                    case DirectiveKind.CodeBlock:
+                        AcceptWhile(IsSpacingToken(includeNewLines: true, includeComments: true));
+                        Output(SpanKindInternal.Markup, AcceptedCharactersInternal.AllWhiteSpace);
 
-                        using (PushSpanConfig())
+                        ParseDirectiveBlock(descriptor, parseChildren: (startingBraceLocation) =>
                         {
-                            HtmlParser.ParseRazorBlock(Tuple.Create("{", "}"), caseSensitive: true);
-                        }
-
-                        Span.Start = CurrentLocation;
-                        Initialize(Span);
-
-                        IsNested = wasNested;
-
-                        NextToken();
-                    });
-                    break;
-                case DirectiveKind.CodeBlock:
-                    AcceptWhile(IsSpacingToken(includeNewLines: true, includeComments: true));
-                    Output(SpanKindInternal.Markup, AcceptedCharactersInternal.AllWhiteSpace);
+                            NextToken();
+                            Balance(BalancingModes.NoErrorOnFailure, CSharpSymbolType.LeftBrace, CSharpSymbolType.RightBrace, startingBraceLocation);
+                            Span.ChunkGenerator = new StatementChunkGenerator();
+                            Output(SpanKindInternal.Code);
+                        });
+                        break;
+                }
+            }
+            finally
+            {
+                if (directiveErrorSink.Errors.Count > 0)
+                {
+                    var directiveDiagnostics = directiveErrorSink.Errors.Select(error => RazorDiagnostic.Create(error));
+                    directiveChunkGenerator.Diagnostics.AddRange(directiveDiagnostics);
```

Comment:
Capturing the diagnostics on the parent directive and not on the actual tokens because we can't properly parse the tokens (they're errored) without risking throwing the entire parsing system into disarray.